src/boot.c

```diff
@@ -412,31 +427,152 @@ static void read_fsinfo(DOS_FS * fs, struct boot_sector *b, unsigned int lss)
+}
+
+static int try_boot_sector(unsigned *boot_sector, unsigned offset,
+                           struct boot_sector *b)
+{
+    unsigned int logical_sector_size;
+    off_t byte_offset = offset * SECTOR_SIZE;
+
+    fs_read(byte_offset, sizeof(*b), b);
+    logical_sector_size = GET_UNALIGNED_W(b->sector_size);
+
+    if (
+	!logical_sector_size ||
+	/* Logical sector size is zero. */
+
+	/* This is the first thing that will fail if the platform needs special
+	 * handling of unaligned multibyte accesses but such handling isn't
+	 * being provided. See GET_UNALIGNED_W() above. */
+	logical_sector_size & (SECTOR_SIZE - 1) ||
+	/* Logical sector size is not a multiple of the physical sector size. */
+
+	!b->cluster_size ||
+	/* Cluster size is zero. */
+
+	!b->fats ||
+	/* Number of FATs is zero. (NTFS is like this.) */
+
+	byte_offset % logical_sector_size
+	/* ...else the boot sector isn't on a logical sector boundary. */
+	)
+    {
+	return 0;
+    }
+
+    *boot_sector = byte_offset / logical_sector_size;
+    return 1;
+}
+
+static int has_signature(const void *sector)
+{
+    const uint8_t *sector_u8 = sector;
+    return sector_u8[510] == 0x55 && sector_u8[511] == 0xaa;
+}
+
+static int try_backup_boot_sector(unsigned *boot_sector, unsigned offset,
+                                  struct boot_sector *b)
+{
+    /* Backup boot sectors are for FAT32 only. Also, check backup boot sectors
+     * for a signature. */
+    int success =
+	try_boot_sector(boot_sector, offset, b) &&
+	is_fat32(b) &&
+	has_signature(b);
+
+    if (success) {
+	unsigned short sectors = GET_UNALIGNED_W(b->sectors);
+
+	printf(
+	    "Candidate backup boot sector found at sector %u:\n"
+	    "%10d bytes per logical sector\n"
+	    "%10u sectors total\n"
+	    "Use this backup boot sector?\n",
+	    *boot_sector,
+	    GET_UNALIGNED_W(b->sector_size),
+	    sectors ? sectors : le32toh(b->total_sect));
+	return get_choice(1, "  Using this backup boot sector.",
+			  2,
+			  1, "Use this backup boot sector",
+			  2, "Search for a different backup boot sector") == 1 ?
+	    1 :
+	    0;
+    }
+
+    return success;
+}
+
 void read_boot(DOS_FS * fs)
 {
     struct boot_sector b;
+    unsigned boot_sector;
     unsigned total_sectors;
     unsigned int logical_sector_size, sectors;
     long long fat_length;
     unsigned total_fat_entries;
     off_t data_size;
     long long position;
 
-    fs_read(0, sizeof(b), &b);
+    if (!try_boot_sector(&boot_sector, 0, &b)) {
+	printf("Boot sector is unusable.\n");
+
+	/* Backup boot sector is at sector 6 by default. Check that first, for
+	 * valid sector sizes of 512 to 4096 bytes.
+	 *
+	 * Might not be a bad idea to only check this for disks large enough to
+	 * accommodate FAT32. */
+	if (!try_backup_boot_sector(&boot_sector, 6, &b) &&
+	    !try_backup_boot_sector(&boot_sector, 12, &b) &&
+	    !try_backup_boot_sector(&boot_sector, 24, &b) &&
+	    !try_backup_boot_sector(&boot_sector, 48, &b)) {
```

Comment:
Due to how you check for sector size, this code does not do what is written in comment above. Third test also checks for sector 24 with logical sector size 512 bytes. But instead third test should check only for sector 6 with logical sector size 2048 bytes.